eval.go

```diff
@@ -568,30 +568,32 @@ func (d *indexData) gatherBranches(docID uint32, mt matchTree, known map[matchTr
 }
 
 func (d *indexData) List(ctx context.Context, q query.Q, opts *ListOptions) (rl *RepoList, err error) {
-	var include func(rle *RepoListEntry) (bool, error)
+	var include func(rle *RepoListEntry) bool
 
 	q = d.simplify(q)
 	if c, ok := q.(*query.Const); ok {
 		if !c.Value {
 			return &RepoList{}, nil
 		}
-		include = func(rle *RepoListEntry) (bool, error) {
-			return true, nil
+		include = func(rle *RepoListEntry) bool {
+			return true
 		}
 	} else {
-		// We need to run a search per repo to decide if it is included.
-		include = func(rle *RepoListEntry) (bool, error) {
-			qOneRepo := query.NewAnd(
-				query.NewRepoSet(rle.Repository.Name),
-				q)
-			sr, err := d.Search(ctx, qOneRepo, &SearchOptions{
-				ShardMaxMatchCount: 1,
-				TotalMaxMatchCount: 1,
-			})
-			if err != nil {
-				return false, err
-			}
-			return len(sr.Files) > 0, nil
+		sr, err := d.Search(ctx, q, &SearchOptions{
+			ShardRepoMaxMatchCount: 1,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		foundRepos := make(map[uint32]struct{}, len(sr.Files))
+		for _, file := range sr.Files {
+			foundRepos[file.RepositoryID] = struct{}{}
```

Comment:
technically RepositoryID can be unset and in practice it will be for OSS users of zoekt. So unfortunetly we can't use this as a key. Previously we relied on name, this is safer (if not slower).

Reply:
Ah, shucks. Good catch. I've updated this to use repository name here.